dissect/target/plugins/os/unix/log/utmp.py

```diff
@@ -68,6 +122,120 @@ def __iter__(self):
 
         while True:
             try:
-                yield utmp.entry(byte_stream)
+                entry = utmp.entry(byte_stream)
+
+                r_type = ""
+                if entry.ut_type in utmp.Type.reverse:
+                    r_type = utmp.Type.reverse[entry.ut_type]
+
+                ut_host = entry.ut_host.decode(errors="surrogateescape").strip("\x00")
+                ut_addr = None
+
+                # UTMP misuses the field ut_addr_v6 for IPv4 and IPv6 addresses, because of this
+                # the ut_host field is used to determine if the ut_addr_v6 is an IPv6 address
+                # where the last 12 bytes of trailing zeroes.
+                if entry.ut_addr_v6:
+                    if not entry.ut_addr_v6[1:] == [0, 0, 0]:
+                        # IPv6 address that uses > 4 bytes
+                        ut_addr = ipaddress.ip_address(struct.pack("<4i", *entry.ut_addr_v6))
+                    else:
+                        try:
+                            if isinstance(ipaddress.ip_address(ut_host), ipaddress.IPv6Address):
+                                # IPv6 address that uses 4 bytes with 12 bytes of trailing zeroes.
+                                ut_addr = ipaddress.ip_address(struct.pack("<4i", *entry.ut_addr_v6))
+                            elif isinstance(ipaddress.ip_address(ut_host), ipaddress.IPv4Address):
+                                # IPv4 address (ut_host, ut_addr_v6)
+                                ut_addr = ipaddress.ip_address(struct.pack("<i", entry.ut_addr_v6[0]))
+                            else:
+                                pass
+                        except ValueError:
+                            # NOTE: in case the ut_host does not contain a valid IPv6 address,
+                            # ut_addr_v6 is parsed as IPv4 address. This could not lead to incorrect
+                            # results.
+                            ut_addr = ipaddress.ip_address(struct.pack("<i", entry.ut_addr_v6[0]))
+
+                utmp_entry = UTMP_ENTRY(
+                    ts=from_unix(entry.ut_tv.tv_sec),
+                    ut_type=r_type,
+                    ut_pid=entry.ut_pid,
+                    ut_user=entry.ut_user.decode(errors="surrogateescape").strip("\x00"),
+                    ut_line=entry.ut_line.decode(errors="surrogateescape").strip("\x00"),
+                    ut_id=entry.ut_id.decode(errors="surrogateescape").strip("\x00"),
+                    ut_host=ut_host,
+                    ut_addr=ut_addr,
+                )
+
+                yield utmp_entry
             except EOFError:
                 break
+
+
+class UtmpPlugin(Plugin):
+    WTMP_GLOB = "/var/log/wtmp*"
+    BTMP_GLOB = "/var/log/btmp*"
+
+    def check_compatible(self) -> None:
+        if not any(
```

Comment:
In case BSD support is chosen and modification of the c-struct is needed, can this be done in another PR? I would like to use this plugin for Linux host investigations without observing random IPv4 addresses that should be IPv6 addresses.

Reply:
I think doing BSD support in a separate PR is ok. I think if you just check `self.target.os == linux` in `check_compatible()` for now, we won't need to move it to the linux subdir and users on BSD platforms won't get surprises.